### ISSUE TITLE
static File web module doesn't support byte ranges

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11034]
|<img alt="jmg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jmg reported|
|-|-|
|Trac ID|trac#1493|
|Type|enhancement|
|Created|2006-02-18 13:43:36Z|
|Branch|https://github.com/twisted/twisted/tree/static-file-byte-range-1493-3|

Attachments:

* [twisted.web.static.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb7/bb7e0878197e923c50408c8cccedb73e9630a7f8/64a8e0f62a00b333851111f4c7fca6218517a5b1.patch) (4044 bytes) - added by jmg on 2006-02-18 13:43:28Z - 
* [twisted.web.static.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb7/bb7e0878197e923c50408c8cccedb73e9630a7f8/1ae7080240026e34450d6791ca3c4ad4b72b88a9.diff) (7170 bytes) - added by brubelsabs on 2008-03-18 09:08:17Z - Simple Range-Header support with a simple unit-test.
* [twisted.web.static.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb7/bb7e0878197e923c50408c8cccedb73e9630a7f8/448707b89f9e966d4ae1298baf74d10e9c015217.diff) (13434 bytes) - added by brubelsabs on 2008-03-20 12:51:38Z - Due to the review from jknight, some improvments of the previous patch
* [twisted.web.static.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb7/bb7e0878197e923c50408c8cccedb73e9630a7f8/e0ed471b54cb72bdee6b8a863c798ee58f54aac1.diff) (17203 bytes) - added by brubelsabs on 2008-04-21 13:43:55Z - beautified docstrings and testnames as well 2 blanks between methods, and new MalformedHeader Exception derived from Error
* [twisted.web.static.4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb7/bb7e0878197e923c50408c8cccedb73e9630a7f8/0999ac6858590766f64e68913ffa5ed0f2d3365c.diff) (24966 bytes) - added by brubelsabs on 2008-05-13 14:44:14Z - improvements for the suggestions from exarkun

<details><summary>Searchable metadata</summary>

```
trac-id__1493 1493
type__enhancement enhancement
reporter__jmg jmg
priority__normal normal
milestone__ 
branch__branches_static_file_byte_range_1493_3 branches/static-file-byte-range-1493-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1140270216000000 1140270216000000
changetime__1216990519000000 1216990519000000
version__ 
owner__ 
cc__exarkun cc__jknight cc__jmg cc__brubelsabs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jmg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jmg commented|
|-|-|

```
#!html
<pre>
The File class in web/static.py doesn't support byte servering.  The code is
commented out because it is broken.  Attached is a patch that in my testing
fixes it (at least the basic continue an aborted download).
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="jmg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jmg commented|
|-|-|

```
#!html
<pre>
oh, I forgot to say the patch is against Web 0.5.0 that was part of
TwistedSumo-2005-11-06.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
This isn't likely to get applied without a unit test, and since most development
is going towards twisted.web2, a unit test isn't likely to get written unless
you write it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Exarkun is correct.

Also, priority -> feature.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

[#2986](https://github.com/twisted/twisted/issues/2986) was closed as a duplicate of this one.

**Comment by twisted-trac** [comment]
|<img alt="brubelsabs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|brubelsabs commented|
|-|-|

I reused the patch above and added a simple unit test the result is in the [[twisted.web.static.diff](None/twisted.web.static.diff) attachment]

Basically, I think the position of some of this code (start stop vars) belongs more to the FilePath class than to the File class, because in the resumeProducing method the real data is read an written. Additionally the request object is passed to this class too.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_8"></a>|@jyknight commented|
|-|-|

You haven't fixed at least two of the serious issues in the original code:

The parsing is fragile (and then protected via a catch-all except, very bogus), and the range returned is wrong (off-by-one on the endpoint).

**Comment by twisted-trac** [comment]
|<img alt="brubelsabs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|brubelsabs commented|
|-|-|

Thanks jknight! I hope this [[twisted.web.static.2.diff](None/twisted.web.static.2.diff) one] is more done the right way. I moved the chunky test from test_web to test_static, where are already some basic test (which I should have seen earlier ;)). I (hopefully) fixed the off-by-one error, and the except clause catching invalid byte-range-specs is just generic to ValueErrors (handling non integer ranges like "bytes=seom-abc"). So I set it back to review..

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph removed owner|
|-|-|

Thanks for the patch!

  * `SyntaxError` is a scary exception class to use for this; it really means that some Python syntax was invalid.  It should probably be a custom exception that specifically deals with this case.
  * The failure is logged (good) but there's no test for this logging (not good).  You can do this by adding a log observer for the duration of your test.  Also, the `repr()`'d invalid range header (or at least a part of it) should always be included in the log message so that we can see if there is a parsing case we've missed.

The twisted.web code is kind of old, and you followed it consistently which I appreciate, but our coding-standard preferences have shifted since it was originally written:
  * Please put 2 blank lines between methods
  * Test methods should be named `test_fooBar` not `testFooBar`.
  * docstring quotes should be on a line by themselves, i.e.
```
"""
Like this
"""
```
    not
```
"""Like this"""
```

Trac doesn't allow me to assign this to you for some reason so I'm reassigning to nobody...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to brubelsabs|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="brubelsabs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|brubelsabs commented|
|-|-|

Hmm, though I've to improve again and again, I see this [[twisted.web.static.3.diff](None/twisted.web.static.3.diff)    patch] gets better and better :-) So set it back to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Cool, thanks! :)

Here's some more feedback:

  * in `test_web.py`, `FakeRequest.getHeader` has no docstring - could you add one?
  * `FakeRequest` doesn't document any of its instance attributes.  It should document all of them.  Could you add documentation for at least the new ones you're adding?
  * in `test_static.py`, `Range` should have a class docstring, and probably be named `RangeTests` to indicate it's a test case.
  * the `payload` assignment line is wider than 79 cols, it should be split up (you can do this with some parens and string literal concatenation, eg
```python
foo = ('foo'
       'bar')
```
  * Instead of `tempfile.NamedTemporaryFile()`, you may want to use `file(self.mktemp())`.  `mktemp` creates files beneath `_trial_temp` where it's easy to clean them up, or leave them behind for inspection after the test run is done.
  * You might find `self.addCleanup` nicer than `tearDown`.  eg, after opening the temp file, `self.addCleanup(self.file.close)` and after adding the log observer, `self.addCleanup(log.removeObserver, self.catcher.append)`.
  * Also, there's some trailing whitespace in the `setUp` method
  * `_assert_logged` should be `_assertLogged` (or it could be public - there's no API stability guarantees in the test suite)
  * test methods should be `test_foo` rather than `test_Foo`
  * test methods should all have docstrings explaining the case they're testing, for example, `test_BodyLength` might have a docstring like this:
```
"""
A correctly formatted range request consisting of a single range beginning at 0 and ending at a length less than the length of the page is responded to with the substring of the render results defined by that range.
"""
  * "... is correct ..." doesn't need to be in any test docstrings, since it is implied that the test is for correct behavior. ;)
  * There's a bit more trailing whitespace in some of the tests
  * in `error.py`,
    * thanks for adding the docstring to `Error.__init__` :)
    * Can you give `MalformedHeader` a class docstring?
    * there should be direct tests for the cases which raise `MalformedHeader` - have some unit tests that call `File._handleRangeRequest` directly and assert it raises the right exception with the right code
  * in `static.py,
    * `_handleRangeRequest` is a bit long:
      * the `ValueError` it handles covers a lot of code - which failure is this trying to catch?  I guess it's covering all the `int` calls?  Is there a unit test for each of these? (I'm being a bit lazy since I need to go eat - I should be checking to find the answer to that :P)

Thanks for your work on this!  I look forward to the next version of the patch. (when you add the review keyword, re-assign the ticket to nobody so everyone recognizes that they could review it)
```{
"""
A correctly formatted range request consisting of a single range beginning at 0 and ending at a length less than the length of the page is responded to with the substring of the render results defined by that range.
"""
  * "... is correct ..." doesn't need to be in any test docstrings, since it is implied that the test is for correct behavior. ;)
  * There's a bit more trailing whitespace in some of the tests
  * in `error.py`,
    * thanks for adding the docstring to `Error.__init__` :)
    * Can you give `MalformedHeader` a class docstring?
    * there should be direct tests for the cases which raise `MalformedHeader` - have some unit tests that call `File._handleRangeRequest` directly and assert it raises the right exception with the right code
  * in `static.py,
    * `_handleRangeRequest` is a bit long:
      * the `ValueError` it handles covers a lot of code - which failure is this trying to catch?  I guess it's covering all the `int` calls?  Is there a unit test for each of these? (I'm being a bit lazy since I need to go eat - I should be checking to find the answer to that :P)

Thanks for your work on this!  I look forward to the next version of the patch. (when you add the review keyword, re-assign the ticket to nobody so everyone recognizes that they could review it)

**Comment by twisted-trac** [comment]
|<img alt="brubelsabs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|brubelsabs removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [23686]) Branching to 'static-file-byte-range-1493'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [23687]) Apply twisted.web.static.4.diff

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to brubelsabs|
|-|-|

I applied the latest patch to the branch referenced above.  I made some docstring adjustments and other minor changes.

  * twisted/web/test/test_static.py
    * Where did that payload come from?  Is it just random bytes?
    * `test_statusCodeRequestedRangeNotSatisfiable` and `test_invalidStartBytePos` don't seem to be correct based on my reading of the RFC.  I think they are testing syntactically invalid ranges and so should be returning 200, not 416.
    * is `test_multipleRanges` just demonstrating an implementation limitation?  Presumably some future enhancement can support responding to multiple ranges?
  * twisted/web/static.py
    * variables should be named `camelCase` instead of `with_underscores`.
    * It'd probably make sense to do all of the parsing in a separate function from `_doRangeRequest`.

**Comment by twisted-trac** [comment]
|<img alt="brubelsabs's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|brubelsabs commented|
|-|-|

Sorry for beeing so long absent :-/, here the answers as far as I can remember...

 * twisted/web/test/test_static.py
   * Yes the payload is just randomly choosen
   * Yes again, you are right. The RFC 2616 just says, if the current extend is overlapped with the bytes ranges, nothing about invalid byte ranges e.g. 20-13, but if bytes:1023-1024 is requested then if the ressource has not so much bytes, only then 416 should be returned. [AFAIU](http://www.rfc.net/rfc2616.html#s10.4.17)
   * Yes the multipleRanges thing demonstrates an implementation limitation. I though a while over the problem, how one could implement multiple byte ranges, but I think then the class FileTransfer cannot be used that easy way as it is done in the moment. I think this is really hard to implement.

 * twisted/web/static.py
   * Sorry for violating still the coding standards, I hope I can do any further updates
   * Yes _doRangeRequest is grown very big, I remember on some coding styles which told: that a method should fill at maximum one screen ;), and this is near the limit, so you right again ;)

ATM I really struggle on my work here, so I've little time right now. I would really appreaciate if some one could do the "rest". If not it will take some time for me to carry on, but I won't let the work be for nothing. (oh my english is.. well you know already ;))
   
Exarkun, thanks for your review! Maybe next month, I've a little time to do any further work.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks for your work on this so far brubelsabs, this ticket would be in much worse shape without your help. :)  I hope things at work get smoother for you soon.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [24198]) Branching to 'static-file-byte-range-1493-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [24200]) comment about this pile of randomness

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [24201]) change behavior required by test_statusCodeRequestedRangeNotSatisfiable and test_invalidStartBytePos

these are both examples of syntactically invalid byte ranges and should be
treated as any other syntactically invalid range header.

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [24202]) note that really, multiple ranges *should* be supported; this test is just to
make sure the single range behavior works when multiple ranges are present in
the request.

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [24204]) simplify the test suite by directly testing parsing logic; remove some redundant tests; refactor the parsing logic into a separate method; remove some other unused stuff

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [24205]) Branching to 'static-file-byte-range-1493-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Alright, ready for someone else to take a look.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_27"></a>|@jyknight set owner to @exarkun|
|-|-|

Is the comment you modified in test_multipleRanges ("") even correct at all? Where does the RFC say it's acceptable to return a single range when multiple are asked for? It says you can totally ignore the Range header, but I don't see where it says you can ignore parts of it.

test_invalidStartBytePos is testing completely wrong behavior. Firstly, a range with start greater than resource length isn't syntactically invalid, secondly the RFC is clear about what you're supposed to do with an unsatisfiable range request (return a 416 response).

The _parseRangeHeader doesn't follow HTTP RFC parsing rules w.r.t. whitespace. After every split, you need .strip. Or, use a generic tokenizer.

It seems a little silly to have a 3-line log entry whenever someone sends a random bad http header.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

(In [24209]) send 416 when a range starts after the resource ends, fixing test_invalidStartBytePos

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [24210]) just give up on multiple byte ranges instead of pretending there was juist one

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun commented|
|-|-|

(In [24211]) a test for ranges with whitespace (which passes without changes to the implementation, although only by accident)

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_31"></a>|@exarkun commented|
|-|-|

(In [24212]) add a test for null elements in the bytes range and make it pass by ignoring all null elements

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

(In [24213]) shorten the log message for ignored range headers to 1 line

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun removed owner|
|-|-|

Thanks for the review!

I think I addressed everything.  'test_multipleRanges` doesn't exist anymore (unless I've managed to confuse myself), so I interpreted your comment about that function as meaning that I should get rid of `test_multiRange` and add that case to `test_invalidRanges` or that I should implement real multiple range support (I did the former).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun commented|
|-|-|

(In [24215]) Add a couple more whitespace-related range parsing tests

`<foom> exarkun: self.assertEqual(self.resource._parseRangeHeader(' bytes =  1 - 2 '), (1, 2))`

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_35"></a>|@ghazel commented|
|-|-|

Looking at static-file-byte-range-1493-3:

File._parseRangeHeader's docstring says "Return a three-tuple of the type, start, and stop value" but it only returns (start, end).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

(In [24239]) Correct docstring for _parseTangeHeader

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_37"></a>|@therve set owner to @exarkun|
|-|-|

`DummyRequest` has a 2 getHeader methods


``` 
+        range = request.getHeader('range')
+        if range is not None:
```

range is a builtin, another name would be better.


```
+        self.assert_(len(self.catcher) == 0,
```

It should use assertEquals

 
```
+        content_length = stop = self.getFileSize()
```

Please rename it contentLenght.

As all the points are minor, please merge it when corrected. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun commented|
|-|-|

(In [24315]) address review comments

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun set status to `closed`|
|-|-|

(In [24318]) Merge static-file-byte-range-1493-3

Author: brubelsabs, exarkun
Reviewer: exarkun, jknight, ghazel, therve
Fixes: [#11034](https://github.com/twisted/twisted/issues/11034)

Add support for simple Range header requests to `twisted.web.static.File`.
Single byte-range requests will now be honored by that resource.  Other
range requests will be ignored, as before.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun set status to `reopened`|
|-|-|

(In [24322]) Revert r24318 - test suite regression on Windows

A number of tests from twisted.web.test.test_static.RangeTests
fail on Windows with this revision.

Reopens [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_41"></a>|@exarkun commented|
|-|-|

(In [24323]) make sure the file gets closed before we try to look at what's in it, because Windows has very weak guarantees about when you'll actually be able to see your data

refs [#11034](https://github.com/twisted/twisted/issues/11034)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="colin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_43"></a>|colin set owner to @exarkun|
|-|-|

Works for me. Tested with wget and FireFox 3.0.1. 

It does raise a question about how precise logging will be with this change though. Unless the range start is deducted from the file size for CLF or a combined log format, I don't see how any of Apache's logging standards are accurate...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_44"></a>|@exarkun set status to `closed`|
|-|-|

(In [24361]) Merge static-file-byte-range-1493-3

Author: brubelsabs, exarkun
Reviewer: exarkun, jknight, ghazel, therve, colin
Fixes: [#11034](https://github.com/twisted/twisted/issues/11034)

Add support for simple Range header requests to `twisted.web.static.File`.
Single byte-range requests will now be honored by that resource. Other range
requests will be ignored, as before.

This re-merge changes the handling of files in the unit tests so that it is
correct on Windows, where data does necessarily show up in files after it
has been written but before the file has been closed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
